openedx_authz/api/data.py

```diff
@@ -0,0 +1,258 @@
+"""Data classes and enums for representing roles, permissions, and policies."""
+
+from enum import Enum
+from typing import Literal
+
+from attrs import define
+
+
+class GroupingPolicyIndex(Enum):
+    """Index of fields in a grouping policy."""
+
+    SUBJECT = 0
+    ROLE = 1
+    SCOPE = 2
+    # The rest of the fields are optional and can be ignored for now
+
+
+class PolicyIndex(Enum):
+    """Index of fields in a policy."""
+
+    ROLE = 0
+    ACT = 1
+    SCOPE = 2
+    EFFECT = 3
+    # The rest of the fields are optional and can be ignored for now
+
+
+@define
+class AuthZData:
+    """Base class for all authz data classes.
+
+    Attributes:
+        NAMESPACE: The namespace prefix for the data type (e.g., 'user', 'role').
+        SEPARATOR: The separator between the namespace and the identifier (e.g., ':', '@').
+
+    Subclasses are automatically registered by their NAMESPACE for factory pattern.
+    """
+
+    SEPARATOR: str = "@"
+    NAMESPACE: str = None
+
+    # TODO: Implement factory method to return correct subclass based on NAMESPACE prefix.
+    # This would allow initializing with either subject or scope, etc. and returning the correct subclass.
+    # So we don't have to manage each subclass separately or hardcoded anywhere.
+
+
+@define
+class ScopeData(AuthZData):
+    """A scope is a context in which roles and permissions are assigned.
+
+    Attributes:
+        scope_id: The scope identifier (e.g., 'org@Demo').
+
+    Acts as a factory: automatically returns the correct subclass based on the scope_id prefix.
+    """
+
+    NAMESPACE: str = "sc"
+    scope_id: str = ""
+    name: str = ""
+
+    def __attrs_post_init__(self):
+        """Ensure scope ID has appropriate namespace prefix."""
+        if not self.scope_id:
+            self.scope_id = f"{self.NAMESPACE}{self.SEPARATOR}{self.name}".lower()
+
+        # Allow reverse lookup of name from scope_id
+        if (
+            not self.name
+            and self.scope_id
+            and self.NAMESPACE
+            and self.scope_id.startswith(f"{self.NAMESPACE}{self.SEPARATOR}")
+        ):
+            self.name = self.scope_id.split(self.SEPARATOR, 1)[1].lower()
+
+
+@define
+class ContentLibraryData(ScopeData):
```

Comment:
In the future I assume we will want to move these Library specific classes to live with Libraries itself and work as a plugin?

Reply:
Yes! That would be ideally done by the extension mechanism we implement supported by the scope registry: https://github.com/openedx/openedx-authz/pull/75/files#diff-5bd0f6e4e6ab9e8db665a0d717ca3e5867b9e0da168abf24493a3ceac4d99ad5R105-R107